pane: Hide "Copy Relative Path" and "Reveal In Project Panel" actions for files outside of the projects

Poldraunic opened the pull request:
"Copy Relative Path" action had a check for existence of relative path but it always passed because [`WorktreeStore::find_worktree()`](https://github.com/zed-industries/zed/blob/1d5499bee72f1ea57ed883b942f05d5c1e7cec89/crates/project/src/worktree_store.rs#L148) function returned empty path for these kinds of files. It feels correct to make changes there, but I don't know what else could be impacted. 

"Reveal In Project Panel" had no check whatsoever and so I made one.

Release Notes:

- N/A

Poldraunic commented:
Addressing [original comment](https://github.com/zed-industries/zed/pull/23349#pullrequestreview-2583760100:) on original PR:
1. I've removed file existence check altogether because it was not working for all cases. I dunno why I even went with that. I think my test cases we flawed. Now we simply check for empty string which is weird, but seems to cover all cases.
2. I don't think this is the case at all? I was using printf debugging and the code was called once per activation.
3. This is definitely out of my depth and scope of this PR, I think.

SomeoneToIgnore commented on crates/workspace/src/pane.rs:
```diff
@@ -2544,7 +2544,9 @@ impl Pane {
-                            .map(|project_path| project_path.path);
+                            .map(|project_path| project_path.path)
+                            .filter(|path| !path.as_os_str().is_empty());
+                        let has_relative_path = relative_path.is_some();
```
```suggestion
                        let visible_in_project_panel = relative_path.is_some()
                            && worktree.is_some_and(|worktree| worktree.read(cx).is_visible());
```

We could do
```rs
let project_path = pane
    .read(cx)
    .item_for_entry(entry, cx)
    .and_then(|item| item.project_path(cx));
let worktree = project_path.as_ref().and_then(|project_path| {
    pane.read(cx)
        .project
        .upgrade()?
        .read(cx)
        .worktree_for_id(project_path.worktree_id, cx)
});
```

a bit above this and check that we're displaying anything in the panel to reveal.

SomeoneToIgnore commented on crates/workspace/src/pane.rs:
```diff
@@ -2544,7 +2544,9 @@ impl Pane {
                             .and_then(|item| item.project_path(cx))
-                            .map(|project_path| project_path.path);
+                            .map(|project_path| project_path.path)
+                            .filter(|path| !path.as_os_str().is_empty());
```
Zed's workspaces can either "start" with the directory or file in the root entry, and the latter won't have any relative paths, only the workspace root pointing at the file.

This check works, and might be ok as some other places use it, but we can also do 
```rs
// see the comment below to show where worktree is retrieved from
let has_relative_path = worktree
    .read(cx)
    .root_entry()
    .map_or(false, |entry| entry.is_dir());
```

and filter `relative_path` by this instead of `.is_empty()`.

While slightly longer, I think it's less magical this way?

(we sure have to improve the type system to restrict getting "" paths so simply)

SomeoneToIgnore reviewed (commented):
Thank you, this looks more viable as relies on in-memory data instead of FS.

Let's do a bit more descriptive checks and merge this.

Poldraunic commented:
Done.

Poldraunic commented on crates/workspace/src/pane.rs:
```diff
@@ -2566,6 +2566,24 @@ impl Pane {
+                                .read(cx)
+                                .worktree_for_id(project_path.worktree_id, cx)
+                        });
+                        let has_relative_path = worktree.clone().is_some_and(|worktree| {
```
Is `clone()` okay here? 

I mean in a sense of I could've done `is_some()` and then `unwrap()` to do the same thing, but that wouldn't seem like Rust way to do things. And I have no idea how expensive `clone()` is.

SomeoneToIgnore replied:
`worktree` here is an `Entity<Worktree>` which is a fat pointer similar to `Arc`, so not that expensive, yet we can do without it, it seems? 

```suggestion
                        let has_relative_path = worktree.as_ref().is_some_and(|worktree| {
```

SomeoneToIgnore commented on crates/workspace/src/pane.rs:
```diff
@@ -2574,7 +2592,11 @@ impl Pane {
                             .read(cx)
                             .item_for_entry(entry, cx)
                             .and_then(|item| item.project_path(cx))
```
We have `let project_path` doing the same "get a project path" above, so can use it here.

Poldraunic replied:
Good catch. I was thinking about updating it here, but it slipped my mind.

SomeoneToIgnore reviewed (commented):
Thank you, just a few more small nits.

Poldraunic committed:
pane: Hide "Reveal In Project Panel" for files outside of projects

Poldraunic committed:
pane: Hide "Copy Relative Path" for files outside of projects

The check for this action is already in place but it always passes because resolved relative Path contains empty string (as opposed to None).

SomeoneToIgnore reviewed (approved):
Thank you.
